spyder_kernels/customize/code_runner.py

```diff
@@ -0,0 +1,593 @@
+import time
+from functools import partial
+from contextlib import contextmanager
+
+from IPython.core.inputtransformer2 import (
+    TransformerManager,
+    leading_indent,
+    leading_empty_lines,
+)
+from IPython.core.magic import (
+    needs_local_scope,
+    no_var_expand,
+    line_cell_magic,
+    magics_class,
+    Magics,
+)
+
+from spyder_kernels.comms.frontendcomm import frontend_request, CommError
+from spyder_kernels.customize.namespace_manager import NamespaceManager
+from spyder_kernels.customize.spyderpdb import SpyderPdb
+from spyder_kernels.customize.umr import UserModuleReloader
+from spyder_kernels.customize.utils import capture_last_Expr, canonic, create_pathlist
+
+
+__umr__ = UserModuleReloader(namelist=os.environ.get("SPY_UMR_NAMELIST", None))
+logger = logging.getLogger(__name__)
+SHOW_GLOBAL_MSG = True
+SHOW_INVALID_SYNTAX_MSG = True
+
+
+@magics_class
+class SpyderCodeRunner(Magics):
+    """
+    Functions and Magics related to code execution, debugging, profiling, etc.
+    """
+
+    def runfile(self, filename=None, args=None, wdir=None, namespace=None,
+                post_mortem=False, current_namespace=False):
+        """
+        Run a file.
+        """
+        if namespace is None:
+            namespace = self.shell.user_ns
+            local_ns = self.shell.get_local_scope(1)
+        else:
+            local_ns = None
+            current_namespace = True
+        if filename is None:
+            filename = self._get_current_file_name()
+        canonic_filename = canonic(filename)
+        return self._exec_file(
+            filename=filename,
+            canonic_filename=canonic_filename,
+            args=args,
+            wdir=wdir,
+            post_mortem=post_mortem,
+            current_namespace=current_namespace,
+            context_globals=namespace,
+            context_locals=local_ns,
+        )
+
+    def debugfile(self, filename=None, args=None, wdir=None, namespace=None,
+                  post_mortem=False, current_namespace=False):
+        """
+        Debug a file.
+        """
+        if namespace is None:
+            namespace = self.shell.user_ns
+            local_ns = self.shell.get_local_scope(1)
+        else:
+            local_ns = None
+            current_namespace = True
+        if filename is None:
+            filename = self._get_current_file_name()
+        canonic_filename = canonic(filename)
+        with self._debugger_exec(canonic_filename, True) as debug_exec:
+            self._exec_file(
+                filename=filename,
+                canonic_filename=canonic_filename,
+                args=args,
+                wdir=wdir,
+                current_namespace=current_namespace,
+                exec_fun=debug_exec,
+                post_mortem=post_mortem,
+                context_globals=namespace,
+                context_locals=local_ns,
+            )
+
+    def profilefile(self, filename=None, args=None, wdir=None, namespace=None,
```

Comment:
I suggest to remove this method and leave it for another PR because this PR is more about improving and refactoring the way we execute code in the kernel.

That way, we'll be able to merge thisone along with PR https://github.com/spyder-ide/spyder/pull/20726 as a first step to finally merge the Profiler improvements you've been working on.

Reply:
ok